### ISSUE TITLE
trailing slash "incoherence"

### ISSUE THREAD

**Issue by ivan-redooc** [codemix/yii2-localeurls#39]
Hi,
the problem is the generation of the root URL, sometimes it has the trailing slash sometimes not.

I mean "root URL" a URL without pathinfo

In main.php I've set:
```php
            'languages'               => ['it-*', 'en-*',],
            'showScriptName'  => false,
            'enableLanguageDetection'  => false,
            'enableDefaultLanguageUrlCode' => true,
            'enablePrettyUrl' => true,
```

The incoherence redised in two functions
* the function **createUrl** do append the final slash
* the function **redirectToLanguage** don't append the final slash

I think tha the correct beavior is **set the final slash**

### Some examples using createUrl

call |  URL generated
----------------------- | ------------- 
$this->createUrl(['/']) | /it/
$this->createUrl(['']) | /it/

### Some examples using redirectToLanguage
Initial URL | Redirect to URL | behavior
------------------------ | --------------------------- | -------------
www.example.com/site/index | www.example.com/it/site/index | correct
www.example.com/it/ | www.example.com/it/ | correct
www.example.com/it | www.example.com/it | not correct
www.example.com | www.example.com/it | not correct

A possible  solution could be the follows (UlrManager.php:385):
``` php
        $pathInfo = $this->_request->getPathInfo();
        if ($pathInfo || $language) {
            $redirectUrl .= '/'.$pathInfo;
        }
```

**Comment by mikehaertl** [comment]
Not sure if I understand, because you seem to mix two topics here.

1. "Root URLs"
Not sure what you mean by that. `/` is not a valid route. And if you pass an empty string to createUrl(), the Yii manual states:

> If the route is an empty string, the currently requested route will be used;

2. Redirect for different URLs

What exactly does not work or why do you think URLs with/without a trailing slash incorrect? The redirect URLs keep the trailing slash if there is any.

**Comment by ivan-redooc** [comment]
Yep, sorry, I mixed up too much elements and missed the point.

About the empty route (empty string) your note is about **yii\helpers\Url::to()** (as you can see at https://github.com/yiisoft/yii2/blob/master/docs/guide/runtime-routing.md ), while createRoute() return a "/" (when the application is in the document root or is not in a sub-folder).

I hope that using an empty string with createUrl is correct behavior.

The problem is related to the root-url:
* where the browser is sent when arrive the first time on the site by redirectToLanguage
* where the links point creating the root-url  by createUrl

redirectToLanguage send the browser to `/it` 
createUrl("") create the url `/it/` (see UrlManager.php:200 )

I hope to be more clear this time



 







*

**Comment by mikehaertl** [comment]
Hmm, I still can't see, why this should be a problem. If you used a trailing `/` in your URL, you'll also see a `/` in the redirect URL. If not, then there's no `/`. Why do you think this is a problem?

**Comment by ivan-redooc** [comment]
Hi,
what you describe: 

>  If you used a trailing / in your URL, you'll also see a / in the redirect URL. If not, then there's no /

It's not true in case of `'enableDefaultLanguageUrlCode' => true` and the URL is **http://www.example.org*/*** because the browser is redirected to **http://www.example.org*/it***.

So I request an URL with trailing `/` and I got an URL without  trailing `/` 

IMHO `redirectToLanguage` should use `createUrl`. I'll do a pull request with the my proposal, if you don't mind.

**Comment by mikehaertl** [comment]
Please don't - there's good reason, why `redirectToLanguage()` is the way it is now. We can not use `createUrl()` because we don't know the route at that point. 

> So I request an URL with trailing / and I got an URL without trailing /

I'm still not sure, if there's a problem at all. To me this issue is just cosmetic. There's nothing broken - or do you get an error at some point?

**Comment by ivan-redooc** [comment]
No, no error, except I've 2 different URLs (for Google they're) with the same content. 
I know I can fix with the canonical-meta but it's not likely.

I undersdand you reasons about `redirectToLanguage()`, but please give it a chance:

```
    /**
     * Redirect to the current URL with given language code applied
     *
     * @param string $language for back compatibility.
     */
    protected function redirectToLanguage($language)
    {
        $result = parent::parseRequest($this->_request);
        if ($result !== false) {
            list ($route, $params) = $result;
            $_route = [$route] + $params + $_GET;
        } else {
            throw new \yii\web\NotFoundHttpException(Yii::t('yii', 'Page not found.'));
        }
        $redirectUrl = $this->createUrl($_route);
        Yii::$app->getResponse()->redirect($redirectUrl);
        if (YII_ENV_TEST) {
            throw new \yii\base\Exception(\yii\helpers\Url::to($redirectUrl));
        } else {
            Yii::$app->end();
        }
    }
```

**Comment by mikehaertl** [comment]
I've tested your code - about half of the test cases is failing now. So this does not work.

Again, to make this more clear to me: Can you please describe again, what exactly are the cases where you would expect a different output? Please leave away anything related to `createUrl()`. This has nothing to do with the issue.

The real question is: Which redirects are not OK in your opinion? So only give me a list of URLs and the resulting redirect URL that you consider wrong.

**Comment by mikehaertl** [comment]
And as you mentioned duplicate content: Do you suffer from the same problem as described in #24 maybe?

**Comment by ivan-redooc** [comment]
I write before but I mess-up with markdown :( aplogies.

If the URL I set in the browser is http://www.example.org/ 
the browser is redirected to http://www.example.org/it 
I expected to be http://www.example.org/it/

**Comment by mikehaertl** [comment]
Ok. The open question still is: Why do you expect a trailing slash? Even with Yii you never get a trailing slash except for the root URL.

```
Yii::$app->urlManager->createAbsoluteUrl(['site/index']);
// gives: http://www.example.com/

Yii::$app->urlManager->createAbsoluteUrl(['site/login']);
// gives: http://www.example.com/site/login
```

**Comment by mikehaertl** [comment]
I think I finally understand what your problem is: You use a URL like http://www.example.com/it/some/page and now want to create the root URL for that language:

```
$url = Url::to(['site/index']);
```

This now gives you http://www.example.com/it/ with a trailing slash. To be consistent there should not be a trailing slash in this case.

If this is your problem: I've commited a change, that should fix it. Can you help testing it?

**Comment by ivan-redooc** [comment]
It's work fine!
Thank you.

I don't want bother you, and I promise it's last time I push it. This version pass all the tests.

```
    /**
     * Redirect to the current URL with given language code applied
     *
     * @param string $language the language code to add. Can also be empty to not add any language code.
     */
    protected function redirectToLanguage($language)
    {
        $result = parent::parseRequest($this->_request);
        if ($result === false) {
            throw new \yii\web\NotFoundHttpException(Yii::t('yii', 'Page not found.'));
        }
        list ($route, $params) = $result;
        if($language){
            $params[$this->languageParam]=$language;
        }
        $redirectRoute = [$route] + $params + $_GET;
        $redirectUrl = $this->createUrl($redirectRoute);
        Yii::$app->getResponse()->redirect($redirectUrl);
        if (YII_ENV_TEST) {
            // Response::redirect($url) above will call `Url::to()` internally. So to really
            // test for the same final redirect URL here, we need to call Url::to(), too.
            throw new \yii\base\Exception(\yii\helpers\Url::to($redirectUrl));
        } else {
            Yii::$app->end();
        }
    }
```

**Comment by mikehaertl** [comment]
Hmm, OK, I'm convinced. I probably thought, that `parent::parseRequest()` can't work because we could have a URL with a language code. But we "fix" this via `setPathInfo()` in line 232, so we can rely on `parseRequest()` afterwards. And now that the test cases (hopefully) cover even the most esoteric configurations I think we can consider this change to be safe.

Two things left:

1. I've replaced your `[$route] + $params + $_GET` with a simpler `array_unshift($params, $route)`. Why do you think we should also merge the `$_GET` params? They should be exactly the same as `$params` already.

2. Yii has a minor inconsistency when creating a "root URL" for apps that run from a subdirectory, or that have `showScriptName` set to `true`. So the URL you get from `createUrl(['/'])` looks like:
 * `/baseurl/`
 * `/index.php/`

 Whereas other URLs like `createUrl['site/login']` look like:
 * `/baseurl/site/login`
 * `/index.php/site/login`

 I had to update my test cases, because the extension behaved differently and always created URLs without a trailing slash. But this is now an issue with Yii itself. If you agree, I will not fix that, to be consistent with Yii's default behavior.

**Comment by ivan-redooc** [comment]
Hi

1. I found it in  Request::resolve() and the comment `preserve numeric keys` induce me to replicate the code:
'``
    public function resolve()
    {
        $result = Yii::$app->getUrlManager()->parseRequest($this);
        if ($result !== false) {
            list ($route, $params) = $result;
            $_GET = $params + $_GET; // preserve numeric keys

            return [$route, $_GET];
        } else {
            throw new NotFoundHttpException(Yii::t('yii', 'Page not found.'));
        }
    }
  '''

2. I agree, it's always better to be consistent.

Thanks for your help
Ivan

**Comment by mikehaertl** [comment]
> I found it in Request::resolve()

Hmm, ok, then we better also keep it. I'll change it again.

Thanks for your help. Now we finally got rid of some not so nice parts of the code (actually I also never really liked the old implementation - which you can see from the many comments that I added to make it clearer).

**Comment by mikehaertl** [comment]
Fixed it and even included the latest change (https://github.com/yiisoft/yii2/commit/df6f270a0e691253318b787c23d5332b7dc28389). Changes are now available in latest version 1.2.3.

Thanks again!

**Comment by smilemd** [comment]
yeah, i saw :(  looking for another solution
